core/drivers/i2c/atmel_i2c.c

```diff
@@ -0,0 +1,349 @@
+#include <drivers/pinctrl.h>
+#include <io.h>
+#include <kernel/dt_driver.h>
+#include <libfdt.h>
+#include <matrix.h>
+#include <platform_config.h>
+#include <string.h>
+
+#define TWIHS_CR		0x0
+#define TWIHS_CR_SWRST		BIT(7)
+#define TWIHS_CR_SVDIS		BIT(5)
+#define TWIHS_CR_MSEN		BIT(2)
+#define TWIHS_CR_STOP		BIT(1)
+#define TWIHS_CR_START		BIT(0)
+
+#define TWIHS_MMR		0x4
+#define TWIHS_MMR_DADR_SHIFT	16
+#define TWIHS_MMR_DADR_MASK	0x7F
+#define TWIHS_MMR_MREAD		BIT32(12)
+
+#define TWIHS_CWGR		0x10
+#define TWIHS_CWGR_HOLD_SHIFT	24
+#define TWIHS_CWGR_HOLD_MAX	0x1F
+#define TWIHS_CWGR_CKDIV_SHIFT	16
+#define TWIHS_CWGR_CKDIV_MAX	0x7
+#define TWIHS_CWGR_CHDIV_SHIFT	8
+
+#define TWIHS_CKSRC	BIT32(20)
+
+#define TWIHS_SR	0x20
+#define TWIHS_SR_NACK	BIT32(8)
+#define TWIHS_SR_TXRDY	BIT32(2)
+#define TWIHS_SR_RXRDY	BIT32(1)
+#define TWIHS_SR_TXCOMP	BIT32(0)
+
+#define TWIHS_RHR	0x30
+#define TWIHS_THR	0x34
+
+#define TWIHS_WPMR		0xE4
+#define TWIHS_WPMR_WPKEY	SHIFT_U32(0x545749, 8)
+
+#define I2C_BUS_FREQ	400000
+
+struct atmel_i2c {
+	uint32_t sda_hold_time;
+	vaddr_t base;
+	struct clk *clk;
+};
+
+static TEE_Result atmel_i2c_send_one_byte(struct atmel_i2c *i2c, uint8_t byte)
+{
+	uint32_t sr = 0;
+
+	io_write32(i2c->base + TWIHS_THR, byte);
+
+	while (1) {
+		sr = io_read32(i2c->base + TWIHS_SR);
+		if (sr & TWIHS_SR_NACK) {
+			EMSG("Received NACK while writing !");
+			return TEE_ERROR_GENERIC;
+		}
+		if (sr & TWIHS_SR_TXRDY)
+			break;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static void atmel_i2c_wait_txcomp(struct atmel_i2c *i2c)
+{
+	uint32_t sr = 0;
+
+	while (1) {
+		sr = io_read32(i2c->base + TWIHS_SR);
+		if (sr & TWIHS_SR_TXCOMP)
+			return;
+	}
+}
+
+static void atmel_i2c_send_start(struct atmel_i2c *i2c)
+{
+	io_write32(i2c->base + TWIHS_CR, TWIHS_CR_START);
+}
+
+static void atmel_i2c_send_stop(struct atmel_i2c *i2c)
+{
+	io_write32(i2c->base + TWIHS_CR, TWIHS_CR_STOP);
+}
+
+static TEE_Result atmel_i2c_write_data_no_stop(struct i2c_dev *i2c_dev,
+					       const uint8_t *buf, int len)
+{
+	int i = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct atmel_i2c *i2c = i2c_dev->ctrl->priv;
+	uint32_t mmr = SHIFT_U32(i2c_dev->addr, TWIHS_MMR_DADR_SHIFT);
+
+	io_write32(i2c->base + TWIHS_MMR, mmr);
+
+	for (i = 0; i < len; i++) {
+		res = atmel_i2c_send_one_byte(i2c, buf[i]);
+		if (res)
+			return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result atmel_i2c_write_data(struct i2c_dev *i2c_dev,
+				       const uint8_t *buf, int len)
+{
+	struct atmel_i2c *i2c = i2c_dev->ctrl->priv;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	res = atmel_i2c_write_data_no_stop(i2c_dev, buf, len);
+	if (res)
+		return res;
+
+	atmel_i2c_send_stop(i2c);
+	atmel_i2c_wait_txcomp(i2c);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result atmel_i2c_recv_one_byte(struct atmel_i2c *i2c,
+					  uint8_t *byte)
+{
+	uint32_t sr = 0;
+
+	while (1) {
+		sr = io_read32(i2c->base + TWIHS_SR);
+		if (sr & TWIHS_SR_NACK) {
+			EMSG("Received NACK while reading !");
```

Comment:
"I2C received NACK while reading" maybe? (EMSG does not show function name by default so there is no way to know it is I2C if you don't write it in the message)

Reply:
Indeed, it might be more helpful.